Yoast/wordpress-seo
# Clean links refactor

## Context


* The current links architecture seems to be heavily overengineered. This branch should bring all the link table logic to the `src` directory reusing the concepts we already use there.

## Todo

- [x] Cleanup the dependency chain, merge / remove unneeded files
- [x] Restructure / rename existing classes to make the concepts clear.
- [x] Refactor the storage and installer classes to repositories and migrations.
- [x] Refactor the accessible classes to conditionals.
- [x] Refactor the indexing functionality (indexing plus query classes) to use the indexing framework.
- [x] Everything with register hooks becomes an integration.
- [x] Make sure the feature toggle is no longer a toggle for the backend functionality but only for the features built on top of it.
- [x] Create many to one relationship between links and indexables.
- [x] Write unit tests for new classes.

## Summary


This PR can be summarized in the following changelog entry:

* Refactors the internal link indexing to modern standards.
* Adds indexing of images to the link indexing.

## Test instructions

This PR can be tested by following these steps:

* Set the `link_count` column of all indexables to `NULL`
* Make sure you have a number of posts that include a variety of:
  - Internal links ( a elements with a href that points to the same domain )
  - External links ( a elements with a href to another domain )
  - Internal images ( img element with a src that points to the same domain )
  - External images ( img element with a src that points to another domain )
* Run the link indexation tool ( if available, if you made the above posts on this PR it may not be as they would already be indexed )
* You should now see all links, both normal and images, reflected in the `seo_links` table.
* Update a post, changing which links are present.
* These changes should be reflected in the `seo_links` table after saving.
* Check the link count columns in the post overview, they should be accurate for all normal links ( not images ).
* Navigate to the next page, the links should still be accurate.


## UI changes

## Documentation

## Quality assurance


Takes care of a lot of https://github.com/Yoast/wordpress-seo/issues/15161
